liblumen_alloc/src/erts/process/trace/format.rs

```diff
@@ -0,0 +1,153 @@
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+pub fn format(
+    trace: &Trace,
+    f: &mut fmt::Formatter,
+    kind: Term,
+    reason: Term,
+) -> std::io::Result<()> {
+    use termcolor::Ansi;
+
+    let mut wrapper = unsafe { FormatterWrapper::new(f) };
+    let mut ansi = Ansi::new(&mut wrapper);
+    format_write(trace, &mut ansi, kind, reason)
+}
+
+fn format_write<W>(trace: &Trace, out: &mut W, kind: Term, reason: Term) -> std::io::Result<()>
+where
+    W: WriteColor,
+{
+    let mut bold = ColorSpec::new();
+    bold.set_bold(true);
+    let mut underlined = ColorSpec::new();
+    underlined.set_underline(true);
+    let mut yellow = ColorSpec::new();
+    yellow.set_fg(Some(Color::Yellow));
+    let mut green = ColorSpec::new();
+    green.set_fg(Some(Color::Green));
+
+    out.set_color(&bold);
+    writeln!(out, "Backtrace (most recent call last):")?;
+
+    for symbol in trace.iter_symbols().rev() {
+        println!("{:?}", &symbol);
+        let mfa = symbol.module_function_arity();
+        if mfa.is_none() {
+            continue;
+        }
+        let mfa = mfa.unwrap();
+        let filename = symbol.filename();
+        let line = symbol.line();
+
+        out.reset();
+        write!(out, "  File ");
+
+        match filename {
+            Some(f) => {
+                let filename = trim_filename(f);
+                out.set_color(&underlined);
+                write!(out, "{}", filename);
+                out.reset();
+                write!(out, ":");
+                out.set_color(&yellow);
+                write!(out, "{}", line.unwrap_or(0));
+            }
+            None => {
+                out.set_color(&underlined);
+                write!(out, "<unknown>");
+            }
+        }
+
+        out.reset();
+        write!(out, ", in ");
+        out.set_color(&green);
+        writeln!(out, "{}", &mfa);
+    }
+
+    out.set_color(&bold);
+    let kind: Result<Atom, _> = kind.decode().unwrap().try_into();
+    if let Ok(kind) = kind {
+        if kind == "error" {
+            writeln!(out, "\nProcess raised an error.");
+        } else if kind == "exit" {
+            writeln!(out, "\nProcess exited abnormally.");
+        } else if kind == "throw" {
+            writeln!(out, "\nProcess threw an exception.");
+        } else {
+            writeln!(out, "\nProcess crashed.");
+        }
+    }
+    out.set_color(&yellow);
+    writeln!(out, "  {}\n", reason);
+    out.reset();
+
+    Ok(())
+}
+
+fn trim_filename(file: &Path) -> Cow<'_, str> {
+    let filename = file.to_string_lossy();
+    if filename.starts_with("/rustc/") {
+        if let Some(filename) = filename.get(48..) {
+            Cow::Owned(format!("rust:{}", filename))
+        } else {
+            filename
+        }
+    } else if let Some(basename) = file.file_name().and_then(|x| x.to_str()) {
+        if basename.starts_with('<') && basename.ends_with('>') {
```

Comment:
What's the significance of `<FILENAME>`?

Reply:
That's the format we give virtual files, e.g. `<nofile>`